### DISCUSSION TITLE
Provide request context to custom Adapter

### DISCUSSION THREAD

**Discussion by yanickrochon** [nextauthjs/next-auth#4792] [Ideas]
### Description 📓

While implementing a custom adapter, I can't find any way to obtain the `request` context in order to save device information when performing requests. We use this data in order to debug when clients have issues, so we can track what devices they use to connect, etc. I am aware that I could go around this by using other methods during the authentication pipeline, but it seems redundant to make two requests instead of one.

As far as I can see, all methods from the `Adapter` interface have a single parameter, adding the request as the second argument would provide that functionality.

Ex:

```js
class CustomAdapter extends Adapter {
  useVerificationToken({ identifier, token }, req) {
    // req.clientAddress
  }
}
```

### How to reproduce ☕️

Nothing to reproduce

### Contributing 🙌🏽

Yes, I am willing to help implement this feature in a PR

**Comment by balazsorban44** [comment]
If you mean the incoming `NextApiRequest` you can already access it through https://next-auth.js.org/configuration/initialization#advanced-initialization

I don't see this as a general enough use case to pass it to all adapter methods.

**Reply by yanickrochon** [reply]
So, instead of 

```js
class CustomAdapter extends Adapter {
  useVerificationToken({ identifier, token }, req) {
    // req.clientAddress
  }
}

export default NextAuth({
  ...
  adapter: new CustomAdapter(),
  ....
});
```
which initializes everything once during the application's lifetime, you are suggesting this
```js
class CustomAdapter extends Adapter {
  constructor({ req }) {
    this.req = req;
  }
  useVerificationToken({ identifier, token }) {
    // this.req.clientAddress
  }
}

export default async function auth(req, res) {
  return await NextAuth(req, res, {
    ...
    adapter: new CustomAdapter({ req }),
    ....
  });
}
```
which will get executed at _every single_ requests?

**Comment by yanickrochon** [comment]
The `CredentialProvider` does pass the request context to it's `authorize` callback :

```js
CredentialsProvider({
  name: "Credentials",
  async authorize(credentials, req) {
    const context = getContext(req);

    // ... validate credentials with the given request context
  }
})
```

So why doesn't the `EmailProvider` also pass the request context to it's methods? Why does it require a different method to access that object?

Seriously, it's a dumb fix! Just add the `params` to the adapter methods!

https://github.com/nextauthjs/next-auth/blob/2469e44572f23f709fa8c5c65c6b7a4eb2383e9f/packages/next-auth/src/core/routes/callback.ts#L208-L211

Would become

```js
    const invite = await useVerificationToken?.({
      identifier,
      token: hashToken(token, options),
    }, params)
```

etc.

**Comment by yanickrochon** [comment]
I think this discussion has become stale in favor if this one : #4865
